hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/CompactInternalGenericRecord.java

```diff
@@ -0,0 +1,715 @@
+
+    @Nullable
+    public Class getAssociatedClass() {
+        return associatedClass;
+    }
+
+    public BufferObjectDataInput getIn() {
+        return in;
+    }
+
+    @Override
+    public Schema getSchema() {
+        return schema;
+    }
+
+    @Override
+    @Nonnull
+    public GenericRecordBuilder newBuilder() {
+        return serializer.createGenericRecordBuilder(schema);
+    }
+
+    @Override
+    @Nonnull
+    public GenericRecordBuilder cloneWithBuilder() {
+        return serializer.createGenericRecordCloner(schema, this);
+    }
+
+    @Override
+    @Nonnull
+    public FieldType getFieldType(@Nonnull String fieldName) {
+        return schema.getField(fieldName).getType();
+    }
+
+    @Override
+    public boolean hasField(@Nonnull String fieldName) {
+        return schema.hasField(fieldName);
+    }
+
+    @Nonnull
+    @Override
+    public Set<String> getFieldNames() {
+        return schema.getFieldNames();
+    }
+
+    @Override
+    public byte getByte(@Nonnull String fieldName) {
+        try {
+            return in.readByte(readPrimitivePosition(fieldName, BYTE));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    boolean isFieldExists(@Nonnull String fieldName, @Nonnull FieldType type) {
+        FieldDescriptor field = schema.getField(fieldName);
+        if (field == null) {
+            return false;
+        }
+        return field.getType().equals(type);
+    }
+
+    @Override
+    public short getShort(@Nonnull String fieldName) {
+        try {
+            return in.readShort(readPrimitivePosition(fieldName, SHORT));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public int getInt(@Nonnull String fieldName) {
+        try {
+            return in.readInt(readPrimitivePosition(fieldName, INT));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public long getLong(@Nonnull String fieldName) {
+        try {
+            return in.readLong(readPrimitivePosition(fieldName, LONG));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public float getFloat(@Nonnull String fieldName) {
+        try {
+            return in.readFloat(readPrimitivePosition(fieldName, FLOAT));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public double getDouble(@Nonnull String fieldName) {
+        try {
+            return in.readDouble(readPrimitivePosition(fieldName, DOUBLE));
+        } catch (IOException e) {
+            throw illegalStateException(e);
+        }
+    }
+
+    @Override
+    public boolean getBoolean(@Nonnull String fieldName) {
+        try {
+            FieldDescriptor fd = getFieldDefinition(fieldName, BOOLEAN);
+            int booleanOffset = fd.getOffset();
```

Comment:
seems like you used the offset to mean `number of bits` rather than it is used as `number of bytes` for other field types. this was unexpected to me. ideally, i would like to see a descriptor that has the `booleanOffsetInBytes` and `booleanOffsetWithinLastByte` already exist in the definition and i don not have to make the same static calculation for each read operation. The lines to calculate these can be put into field definition compilation and hence no CPU cycle is wasted.

Reply:
I can do that. Note that this will be part of the serialized field descriptor as well. 
I would keep offset and add just new field bitOffset to the FieldDescriptor where it has `booleanOffsetWithinLastByte` in bits. It will be meaningless for others. I will update implementation accordingly.

Reply:
Fixed by [`1ed047d` (#18790)](https://github.com/hazelcast/hazelcast/pull/18790/commits/1ed047ddbe1e55a9dae2b58e5be09337b47510be)